cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -555,6 +556,94 @@ public void programSpecification(HttpRequest request, HttpResponder responder,
+    ProgramType programType = getProgramType(type);
+    if (programType == null) {
+      throw new MethodNotAllowedException(request.getMethod(), request.getUri());
+    }
+    try {
+      // we are decoding the body to Map<String, String> instead of Map<String, LogEntry.Level> here since Gson will
+      // serialize invalid enum values to null, which is allowed for log level, instead of throw an Exception.
+      lifecycleService.updateProgramLogLevels(
+        new ApplicationId(namespace, appName, appVersion).program(programType, programName), decodeArguments(request));
+      responder.sendStatus(HttpResponseStatus.OK);
+    } catch (JsonSyntaxException e) {
+      responder.sendString(HttpResponseStatus.BAD_REQUEST, "Invalid JSON in body");
+    } catch (SecurityException e) {
+      responder.sendStatus(HttpResponseStatus.UNAUTHORIZED);
+    }
+  }
+
+  /**
+   * Reset the log level for a program back to where it starts. Currently supported program types are
+   * {@link ProgramType#FLOW}, {@link ProgramType#SERVICE} and {@link ProgramType#WORKER}.
+   * The request body can either be empty, which will reset all loggers for the program, or contain a list of
+   * logger names, which will reset for these particular logger names for the program.
+   */
+  @Path("/apps/{app-name}/{program-type}/{program-name}/resetLogLevels")
+  @PUT
+  public void resetProgramLogLevels(HttpRequest request, HttpResponder responder,
+                                    @PathParam("namespace-id") String namespace,
+                                    @PathParam("app-name") String appName,
+                                    @PathParam("program-type") String type,
+                                    @PathParam("program-name") String programName) throws Exception {
+    resetProgramLogLevels(request, responder, namespace, appName, ApplicationId.DEFAULT_VERSION, type, programName);
+  }
+
+  /**
+   * Reset the log level for a program back to where it starts.
+   */
+  @Path("/apps/{app-name}/versions/{app-version}/{program-type}/{program-name}/resetLogLevels")
+  @PUT
+  public void resetProgramLogLevels(HttpRequest request, HttpResponder responder,
+                                    @PathParam("namespace-id") String namespace,
+                                    @PathParam("app-name") String appName,
+                                    @PathParam("app-version") String appVersion,
+                                    @PathParam("program-type") String type,
+                                    @PathParam("program-name") String programName) throws Exception {
+    ProgramType programType = getProgramType(type);
+    if (programType == null) {
+      throw new MethodNotAllowedException(request.getMethod(), request.getUri());
+    }
+    try {
+      Map<String, String[]> resetRequest = parseBody(request, STRING_STRINGARRAY_MAP_TYPE);
+      String[] loggerNames = resetRequest == null ? null : resetRequest.entrySet().iterator().next().getValue();
```

Comment:
This is a really strange request body. The body should be well formed instead of some map with a key that doesn't matter and a list as its value. This is also not guaranteed to be deterministic. If I passed in a body with two keys and two lists, this call may do different things.

Instead, you should enforce that the body is just the list of loggers, or its an object with the loggers as a key. I think the second is better as it would let us evolve the API is needed. something like:
```
{
  "loggers": [ "logger1", "logger2" ]
}
```